sensing/autoware_cuda_pointcloud_preprocessor/src/cuda_pointcloud_preprocessor/cuda_pointcloud_preprocessor_node.cpp

```diff
@@ -245,85 +255,170 @@ void CudaPointcloudPreprocessorNode::imuCallback(
+{
   sensor_msgs::PointCloud2ConstIterator<std::uint32_t> iter_stamp(
     *input_pointcloud_msg_ptr, "time_stamp");
-
-  /* *INDENT-OFF* */
   auto num_points = input_pointcloud_msg_ptr->width * input_pointcloud_msg_ptr->height;
   std::uint32_t first_point_rel_stamp = num_points > 0 ? *iter_stamp : 0;
-  double first_point_stamp = input_pointcloud_msg_ptr->header.stamp.sec +
-                             input_pointcloud_msg_ptr->header.stamp.nanosec * 1e-9 +
-                             first_point_rel_stamp * 1e-9;
+  return input_pointcloud_msg_ptr->header.stamp.sec +
+         input_pointcloud_msg_ptr->header.stamp.nanosec * 1e-9 + first_point_rel_stamp * 1e-9;
+}
 
+void CudaPointcloudPreprocessorNode::updateTwistQueue(double first_point_stamp)
+{
   std::vector<geometry_msgs::msg::TwistWithCovarianceStamped::ConstSharedPtr> twist_msgs =
-    twist_sub_->takeData();
+    twist_sub_->take_data();
   for (const auto & msg : twist_msgs) {
     twistCallback(msg);
   }
   while (twist_queue_.size() > 1 &&
          rclcpp::Time(twist_queue_.front().header.stamp).seconds() < first_point_stamp) {
     twist_queue_.pop_front();
   }
+}
 
-  if (imu_sub_ != nullptr) {
-    std::vector<sensor_msgs::msg::Imu::ConstSharedPtr> imu_msgs = imu_sub_->takeData();
-    for (const auto & msg : imu_msgs) {
-      imuCallback(msg);
-    }
-  }
+void CudaPointcloudPreprocessorNode::updateImuQueue(double first_point_stamp)
+{
+  if (!imu_sub_) return;
 
+  std::vector<sensor_msgs::msg::Imu::ConstSharedPtr> imu_msgs = imu_sub_->take_data();
+  for (const auto & msg : imu_msgs) {
+    imuCallback(msg);
+  }
   while (angular_velocity_queue_.size() > 1 &&
          rclcpp::Time(angular_velocity_queue_.front().header.stamp).seconds() < first_point_stamp) {
     angular_velocity_queue_.pop_front();
   }
-  /* *INDENT-ON* */
-
-  // Obtain the base link to input pointcloud transform
-  geometry_msgs::msg::TransformStamped transform_msg;
+}
 
+std::optional<geometry_msgs::msg::TransformStamped>
+CudaPointcloudPreprocessorNode::lookupTransformToBase(const std::string & source_frame)
+{
   try {
-    transform_msg = tf2_buffer_.lookupTransform(
-      base_frame_, input_pointcloud_msg_ptr->header.frame_id, tf2::TimePointZero);
+    return tf2_buffer_.lookupTransform(base_frame_, source_frame, tf2::TimePointZero);
   } catch (const tf2::TransformException & ex) {
     RCLCPP_WARN(get_logger(), "%s", ex.what());
-    return;
+    return std::nullopt;
   }
+}
+
+std::unique_ptr<cuda_blackboard::CudaPointCloud2> CudaPointcloudPreprocessorNode::processPointcloud(
+  const sensor_msgs::msg::PointCloud2::ConstSharedPtr & input_pointcloud_msg_ptr,
+  const geometry_msgs::msg::TransformStamped & transform_msg)
+{
+  sensor_msgs::PointCloud2ConstIterator<std::uint32_t> iter_stamp(
+    *input_pointcloud_msg_ptr, "time_stamp");
+  std::uint32_t first_point_rel_stamp =
+    input_pointcloud_msg_ptr->width * input_pointcloud_msg_ptr->height > 0 ? *iter_stamp : 0;
```

Comment:
This looks like `getFirstPointTimestamp` from above could be used, or the timestamp could be passed in as a param to this function.